sasmodels/kernel.py

```diff
@@ -40,10 +40,118 @@ class Kernel(object):
+        $F(q, x_k)$, weight $w_k$ and volume $V_k$.  The result is:
+
+        .. math::
+
+            P(q) = \frac{\sum w_k F^2(q, x_k) / \sum w_k}{\sum w_k V_k / \sum w_k}
+
+        The form factor itself is scaled by volume and contrast to compute the
+        total scattering.  This is then squared, and the volume weighted
+        F^2 is then normalized by volume F.  For a given density, the number
+        of scattering centers is assumed to scale linearly with volume.  Later
+        scaling the resulting $P(q)$ by the volume fraction of particles
+        gives the total scattering on an absolute scale. Most models
+        incorporate the volume fraction into the overall scale parameter.  An
+        exception is vesicle, which includes the volume fraction parameter in
+        the model itself, scaling $F$ by $\surd V_f$ so that the math for
+        the beta approximation works out.
+
+        By scaling $P(q)$ by total weight $\sum w_k$, there is no need to make
+        sure that the polydisperisity distributions normalize to one.  In
+        particular, any distibution values $x_k$ outside the valid domain
+        of $F$ will not be included, and the distribution will be implicitly
+        truncated.  This is controlled by the parameter limits defined in the
+        model (which truncate the distribution before calling the kernel) as
+        well as any region excluded using the *INVALID* macro defined within
+        the model itself.
+
+        The volume used in the polydispersity calculation is the form volume
+        for solid objects or the shell volume for hollow objects.  Shell
+        volume should be used within $F$ so that the normalizing scale
+        represents the volume fraction of the shell rather than the entire
+        form.  This corresponds to the volume fraction of shell-forming
+        material added to the solvent.
+
+        The calculation of $S$ requires the effective radius and the
+        volume fraction of the particles.  The model can have several
+        different ways to compute effective radius, with the
+        *effective_radius_type* parameter used to select amongst them.  The
+        volume fraction of particles should be determined from the total
+        volume fraction of the form, not just the shell volume fraction.
+        This makes a difference for hollow shapes, which need to scale
+        the volume fraction by the returned volume ratio when computing $S$.
+        For solid objects, the shell volume is set to the form volume so
+        this scale factor evaluates to one and so can be used for both
+        hollow and solid shapes.
+        """
+        self._call_kernel(call_details, values, cutoff, magnetic, effective_radius_type)
+        #print("returned",self.q_input.q, self.result)
+        nout = 2 if self.info.have_Fq and self.dim == '1d' else 1
+        total_weight = self.result[nout*self.q_input.nq + 0]
+        if total_weight == 0.:
```

Comment:
comparing floats? is this safe? maybe numpy.isclose() instead

Reply:
or try..catch for ZeroDivisionError, since you're ony using shell_volume as a denominator

Reply:
Okay in this case since total weight is the sum of values greater than zero.  I'll add a comment to that effect.